### PR TITLE
feat: no yeoman

### LINKED ISSUE #1152: Default `.gitignore` content generated using yarn 2+ is missing recommended lines by yarn
Do you want to request a *feature* or report a *bug*?
-----------------------------------------------------
A bug.

What is the current behavior?
-----------------------------
OS Version: `Ubuntu 22.04.2 LTS`.

Node version: `v18.16.0` (npm `v9.5.1`, yarn `v3.6.0`).

Oclif version: `v3.9.1`.

Generate a new project using `npx oclif generate <project-name>` with `yarn` as package manager. Yarn configuration shipped with oclif doesn't support [zero-installs](https://yarnpkg.com/features/zero-installs) (it uses `node_modules` local folder). The default `.gitignore` shipped with oclif doesn't follow the [recommended settings by yarn](https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored) (see the second code block).

`.gitignore` default content
```
**/.DS_Store
*-debug.log
*-error.log
/.idea
/.nyc_output
/dist
/lib
/package-lock.json
/tmp
node_modules
oclif.manifest.json
```

What is the expected behavior?
------------------------------
The `.gitignore` file generated when using yarn 2+ should contains all recommended lines from yarn.

`.gitignore` proposed content
```
**/.DS_Store
*-debug.log
*-error.log
/.idea
/.nyc_output
/dist
/lib
/package-lock.json
/tmp
node_modules
oclif.manifest.json
.pnp.*
.yarn/*
!.yarn/patches
!.yarn/plugins
!.yarn/releases
!.yarn/sdks
!.yarn/versions
```

### LINKED ISSUE #1294: Oclif using deprecated transient dependencies and showing npm WARN
Do you want to request a *feature* or report a *bug*?
-----------------------------------------------------
bug

What is the current behavior?
-----------------------------

Using the latest oclif version `4.4.11` there are some deprecated transient dependencies that will show warning messages during running `npm install` of the package.

**Steps to reproduce:**
1. try installing any published CLI library based on `oclif` `4.4.11`
2. the `npm i` will output the following deprecated dependency warnings:
```
npm WARN deprecated debuglog@1.0.1: Package no longer supported. Contact Support at https://www.npmjs.com/support for more info.
npm WARN deprecated readdir-scoped-modules@1.1.0: This functionality has been moved to @npmcli/fs
npm WARN deprecated @npmcli/move-file@1.1.2: This functionality has been moved to @npmcli/fs
npm WARN deprecated @npmcli/move-file@2.0.1: This functionality has been moved to @npmcli/fs
npm WARN deprecated @npmcli/move-file@2.0.1: This functionality has been moved to @npmcli/fs
```

These dependencies originate from `oclif` `4.4.11`:

`debuglog@1.0.1`

```
└─┬ oclif@4.4.11
  └─┬ yeoman-environment@3.19.3
    └─┬ @npmcli/arborist@4.3.1
      └─┬ readdir-scoped-modules@1.1.0
        └── debuglog@1.0.1
```

`readdir-scoped-modules@1.1.0`

```
└─┬ oclif@4.4.11
  └─┬ yeoman-environment@3.19.3
    └─┬ @npmcli/arborist@4.3.1
      └── readdir-scoped-modules@1.1.0
```

`@npmcli/move-file@1.1.2` 

```
└─┬ oclif@4.4.11
  └─┬ yeoman-environment@3.19.3
    └─┬ @npmcli/arborist@4.3.1
      ├── @npmcli/move-file@1.1.2
      └─┬ cacache@15.3.0
        └── @npmcli/move-file@1.1.2 deduped
```

So the main issue is that any CLI npm package based on the Oclif framework using this version will show these warnings during install, which is a pretty bad impression for anyone using the CLI package.

What is the expected behavior?
------------------------------

The expected behaviour is that there are no deprecated transient dependencies showing as npm WARN when installing the package.

Please mention your oclif and OS version.
------------------------------

Oclif version is 4.4.11
OS version is macOS Sonoma 14.0

### PR DESCRIPTION
**Pull request by mdonnalley**
- Removes dependencies on yeoman-generator and yeoman-environment (fixes #1294)
- Exposes all prompts on `oclif generate` as flags so that users can programmatically answer prompts (similar to `npm init --name my-pkg`
- Adds a `--yes` flag on `oclif generate` that will automatically accept defaults for all prompts (similar to `npm init --yes`)
- Adds yarn-specific items to generated .gitignore (fixes #1152)
- Adds `--commands-dir` flag to `oclif generate command` to allow users to select an alternative destination for the generated command (defaults to `src/commands`)
- Standardizes help content across all commands (capitalize first letter, end with period, separate summary and description)

@W-15044742@
@W-15161957@

### CODE REVIEW

**Comment by jpshackelford** [comment]
Very glad to see this work. I am making a choice between oclif and commander today and seeing these warnings had me worried about whether oclif was actively maintained. Thanks for the PR!

**Comment by moltar** [comment]
Can't wait for this to merge. Such an eye sore in the terminal to see all those warnings.

**Comment by mdonnalley** [comment]
Thanks @moltar - hoping to get this out tomorrow or early next week. Just waiting on a coworker to review and test.

File: src/commands/generate.ts [hunk 1]
```diff
@@ -1,26 +1,260 @@
...
+  const headers = {
+    Accept: 'application/vnd.github.v3+json',
+    Authorization: `Bearer ${token}`,
+  }
+
+  try {
+    const {login, name} = await got('https://api.github.com/user', {headers}).json<{login: string; name: string}>()
+    return {login, name}
+  } catch {}
+}
```
Comment (cristiand391) [inline src/commands/generate.ts:25]: I couldn't find where yeoman was using this env, seems it did a while ago but now it does an unauthenticated request, and relies on the user having an email setup in git:

https://github.com/yeoman/generator/blob/71dcba00427d7e382c1c32af68dbb7666ee0269f/src/actions/user.ts#L51-L52

https://github.com/sindresorhus/github-username/blob/c716cf8e920430dc319c44ab7586ab2e72b91467/index.js#L18

Either way works so it should be fine, the precedence seems for getting the username looks like:
1) GH API (if token)
2) git email (if available)
3) pjson.author in template

I'll check on QA 👍🏼

**Review by cristiand391 — APPROVED** [review]
lgtm. I'll QA mainly on windows to ensure the generator works there.

**Comment by cristiand391** [comment]
**QA notes**:

✅ can generate a CLI/command/hook successfully on windows, all paths are correct and can run command/hooks/tests
✅ values provided via flags are respected and only get prompted for the missing ones
✅ both npm/yarn work for generated plugins
NOTE: now that the default PM is `npm` the first `npm install` takes a bit longer because there's no package-lock.json in the templates repo, so users might get something different from what gets tested in the templates repo.

✅ all defaults with `--all`
❌ can't make it pick up my username from local git

latest `oclif`:
Screenshot 2024-03-18 at 18 20 26

this branch:
Screenshot 2024-03-18 at 18 20 52

```
➜  oclif git:(mdonnalley/no-yeoman) ✗ git config --get user.name
Cristian Dominguez
```
so the homepage/bug fields are wrong in the pjson:
```
...
  "homepage": "https://github.com/oclif/qa2",
  "bugs": "https://github.com/oclif/qa2/issues",
...
`

**Pull request merged**